File: cpp/22_GenerateParentheses.cpp
class Solution {
public:
    vector<string> parentheses;
    vector<string> generateParenthesis(int n) {
        helper("", 0, 0, n);
        return parentheses;
    }
    
    void helper(string curr, int left, int right, int n) {
        if (curr.length() == 2 * n) parentheses.push_back(curr);
        else {
            if (left < n) helper(curr + "(", left + 1, right, n);
            if (right < left) helper(curr + ")", left, right + 1, n);
        }
    }
};

File: swift/22_GenerateParentheses.swift
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        func backtrack(_ currentString: String, _ left: Int, _ right: Int) {
            if currentString.count == 2 * n {
                parentheses.append(currentString)
            } else {
                if left < n {
                    backtrack(currentString + "(", left + 1, right)
                }
                
                if right < left {
                    backtrack(currentString + ")", left, right + 1)
                }
            }
        }
        
        var parentheses = [String]()
        backtrack("", 0, 0)
        return parentheses
    }
}
